src/main/java/org/dataloader/orchestration/CF.java

```diff
@@ -0,0 +1,157 @@
+package org.dataloader.orchestration;
+
+import java.util.Iterator;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+public class CF<T> {
+
+    private AtomicReference<Object> result = new AtomicReference<>();
+
+    private static final Object NULL = new Object();
+
+    private static class ExceptionWrapper {
+        private final Throwable throwable;
+
+        private ExceptionWrapper(Throwable throwable) {
+            this.throwable = throwable;
+        }
+    }
+
+    private final LinkedBlockingDeque<CompleteAction<?, ?>> dependedActions = new LinkedBlockingDeque<>();
+
+    private static class CompleteAction<T, V> implements Runnable {
+        Executor executor;
+        CF<V> toComplete;
+        CF<T> src;
+        BiFunction<? super T, Throwable, ? extends V> mapperFn;
+
+        public CompleteAction(CF<V> toComplete, CF<T> src, BiFunction<? super T, Throwable, ? extends V> mapperFn, Executor executor) {
+            this.toComplete = toComplete;
+            this.src = src;
+            this.mapperFn = mapperFn;
+            this.executor = executor;
+        }
+
+        public void execute() {
+            if (executor != null) {
+                executor.execute(this);
+            } else {
+                toComplete.completeViaMapper(mapperFn, src.result.get());
+            }
+
+        }
+
+        @Override
+        public void run() {
+            toComplete.completeViaMapper(mapperFn, src.result.get());
+            src.dependedActions.remove(this);
+        }
+    }
+
+    private CF() {
+
+    }
+
+    public <U> CF<U> newIncomplete() {
+        return new CF<U>();
+    }
+
+    public static <T> CF<T> newComplete(T completed) {
+        CF<T> result = new CF<>();
+        result.encodeAndSetResult(completed);
+        return result;
+    }
+
+    public static <T> CF<T> newExceptionally(Throwable e) {
+        CF<T> result = new CF<>();
+        result.encodeAndSetResult(e);
+        return result;
+    }
+
+    public static <T> CF<T> supplyAsync(Supplier<T> supplier,
+                                        Executor executor) {
+
+        CF<T> result = new CF<>();
+        executor.execute(() -> {
+            try {
+                result.encodeAndSetResult(supplier.get());
+            } catch (Throwable ex) {
+                result.encodeAndSetResult(ex);
+            }
+        });
```

Comment:
Looks a lot like the actual CompletebleFuture code foe scheduling ;) 

`java.util.concurrent.CompletableFuture.AsyncRun`